### ISSUE TITLE
RFC4716 SSH Key stored in AWS does not work on Windows agent for Git clone

### ISSUE THREAD

**Issue by flichtenheld** [jenkinsci/aws-secrets-manager-credentials-provider-plugin#228]
### Jenkins and plugins versions report

<details>
  <summary>Environment</summary>

  
  ```text
  Jenkins: 2.346.1
OS: Linux - 5.15.0-1015-aws
---
ace-editor:1.1
ant:475.vf34069fef73c
antisamy-markup-formatter:2.7
apache-httpcomponents-client-4-api:4.5.13-1.0
artifactory:3.17.0
audit-trail:3.11
authentication-tokens:1.4
aws-credentials:191.vcb_f183ce58b_9
aws-java-sdk:1.12.246-349.v96b_b_f7eb_a_c3c
aws-java-sdk-cloudformation:1.12.246-349.v96b_b_f7eb_a_c3c
aws-java-sdk-codebuild:1.12.246-349.v96b_b_f7eb_a_c3c
aws-java-sdk-ec2:1.12.246-349.v96b_b_f7eb_a_c3c
aws-java-sdk-ecr:1.12.246-349.v96b_b_f7eb_a_c3c
aws-java-sdk-ecs:1.12.246-349.v96b_b_f7eb_a_c3c
aws-java-sdk-elasticbeanstalk:1.12.246-349.v96b_b_f7eb_a_c3c
aws-java-sdk-iam:1.12.246-349.v96b_b_f7eb_a_c3c
aws-java-sdk-logs:1.12.246-349.v96b_b_f7eb_a_c3c
aws-java-sdk-minimal:1.12.246-349.v96b_b_f7eb_a_c3c
aws-java-sdk-sns:1.12.246-349.v96b_b_f7eb_a_c3c
aws-java-sdk-sqs:1.12.246-349.v96b_b_f7eb_a_c3c
aws-java-sdk-ssm:1.12.246-349.v96b_b_f7eb_a_c3c
aws-secrets-manager-credentials-provider:1.2.0
bitbucket:223.vd12f2bca5430
bitbucket-build-status-notifier:1.4.2
blueocean:1.25.5
blueocean-autofavorite:1.2.5
blueocean-bitbucket-pipeline:1.25.5
blueocean-commons:1.25.5
blueocean-config:1.25.5
blueocean-core-js:1.25.5
blueocean-dashboard:1.25.5
blueocean-display-url:2.4.1
blueocean-events:1.25.5
blueocean-git-pipeline:1.25.5
blueocean-github-pipeline:1.25.5
blueocean-i18n:1.25.5
blueocean-jwt:1.25.5
blueocean-personalization:1.25.5
blueocean-pipeline-api-impl:1.25.5
blueocean-pipeline-editor:1.25.5
blueocean-pipeline-scm-api:1.25.5
blueocean-rest:1.25.5
blueocean-rest-impl:1.25.5
blueocean-web:1.25.5
bootstrap4-api:4.6.0-5
bootstrap5-api:5.1.3-7
bouncycastle-api:2.26
branch-api:2.1046.v0ca_37783ecc5
caffeine-api:2.9.3-65.v6a_47d0f4d1fe
checks-api:1.7.4
cloudbees-bitbucket-branch-source:784.v7fcdc7c670f6
cloudbees-folder:6.729.v2b_9d1a_74d673
command-launcher:84.v4a_97f2027398
commons-lang3-api:3.12.0.0
commons-text-api:1.9-9.v39a_53e2e0343
config-file-provider:3.11.1
configuration-as-code:1512.vb_79d418d5fc8
credentials:1139.veb_9579fca_33b_
credentials-binding:523.vd859a_4b_122e6
display-url-api:2.3.6
durable-task:496.va67c6f9eefa7
ec2:1.68
echarts-api:5.3.3-1
email-ext:2.91
external-monitor-job:191.v363d0d1efdf8
favorite:2.4.1
font-awesome-api:6.1.1-1
git:4.11.3
git-client:3.11.0
git-server:1.11
github:1.34.4
github-api:1.303-400.v35c2d8258028
github-branch-source:1677.v731f745ea_0cf
gradle:1.39.4
handlebars:3.0.8
handy-uri-templates-2-api:2.1.8-22.v77d5b_75e6953
htmlpublisher:1.30
ivy:2.2
jackson2-api:2.13.3-285.vc03c0256d517
jakarta-activation-api:2.0.0-3
jakarta-mail-api:2.0.0-6
javadoc:217.v905b_86277a_2a_
javax-activation-api:1.2.0-4
javax-mail-api:1.6.2-7
jaxb:2.3.6-1
jdk-tool:1.5
jenkins-design-language:1.25.5
jjwt-api:0.11.5-77.v646c772fddb_0
jnr-posix-api:3.1.15-1
jquery3-api:3.6.0-4
jsch:0.1.55.2
junit:1119.1121.vc43d0fc45561
ldap:2.10
lockable-resources:2.16
mailer:435.v79ef3972b_5c7
matrix-auth:3.1.5
matrix-project:772.v494f19991984
maven-plugin:3.19
mercurial:2.16.2
mina-sshd-api-common:2.8.0-36.v8e25ce90d4b_1
mina-sshd-api-core:2.8.0-36.v8e25ce90d4b_1
momentjs:1.1.1
multiple-scms:0.8
node-iterator-api:1.5.1
okhttp-api:4.9.3-105.vb96869f8ac3a
pam-auth:1.8
pipeline-build-step:2.18
pipeline-github-lib:38.v445716ea_edda_
pipeline-graph-analysis:195.v5812d95a_a_2f9
pipeline-groovy-lib:598.vcd66b_a_336510
pipeline-input-step:449.v77f0e8b_845c4
pipeline-milestone-step:101.vd572fef9d926
pipeline-model-api:2.2114.v2654ca_721309
pipeline-model-definition:2.2114.v2654ca_721309
pipeline-model-extensions:2.2114.v2654ca_721309
pipeline-rest-api:2.24
pipeline-stage-step:293.v200037eefcd5
pipeline-stage-tags-metadata:2.2114.v2654ca_721309
pipeline-stage-view:2.24
plain-credentials:1.8
plugin-util-api:2.17.0
popper-api:1.16.1-3
popper2-api:2.11.5-2
pubsub-light:1.16
resource-disposer:0.19
saml:2.333.vc81e525974a_c
scm-api:616.ve67136f6c77d
script-security:1175.v4b_d517d6db_f0
slave-setup:1.10
snakeyaml-api:1.30.2-76.vc104f7ce9870
sse-gateway:1.25
ssh-agent:295.v9ca_a_1c7cc3a_a_
ssh-credentials:295.vced876c18eb_4
ssh-slaves:1.821.vd834f8a_c390e
sshd:3.242.va_db_9da_b_26a_c3
structs:324.va_f5d6774f3a_d
throttle-concurrents:2.8
timestamper:1.18
token-macro:308.v4f2b_ed62b_b_16
trilead-api:1.67.vc3938a_35172f
variant:59.vf075fe829ccb
windows-slaves:1.8.1
workflow-aggregator:590.v6a_d052e5a_a_b_5
workflow-api:1188.v0016b_4f29881
workflow-basic-steps:980.v82219a_ed188e
workflow-cps:2759.v87459c4eea_ca_
workflow-cps-global-lib:588.v576c103a_ff86
workflow-durable-task-step:1199.v02b_9244f8064
workflow-job:1207.ve6191ff089f8
workflow-multibranch:716.vc692a_e52371b_
workflow-scm-step:400.v6b_89a_1317c9a_
workflow-step-api:639.v6eca_cd8c04a_a_
workflow-support:833.va_1c71061486b_
ws-cleanup:0.42
  ```

</details>

### What Operating System are you using (both controller, and any agents involved in the problem)?

Jenkins server runs on Ubuntu 20.04 LTS
Windows Agents run Windows Server 2019

### Reproduction steps

Added SSH key as sshUserPrivateKey credential in AWS
Using the key in checkout() pipeline step

### Expected Results

SSH key can be used to clone git repository on every agent

### Actual Results

- git clone works fine on Linux agents
- git clone does not work Windows agents

```text
 > git fetch --tags --force --progress -- git@bitbucket.org:<workspace>/<repo>.git +refs/heads/*:refs/remotes/origin/* # timeout=10
ERROR: Error cloning remote repo 'origin'
hudson.plugins.git.GitException: Command "git fetch --tags --force --progress -- git@bitbucket.org:<workspace>/<repo>.git +refs/heads/*:refs/remotes/origin/*" returned status code 128:
stdout: 
stderr: Load key "C:\\Jenkins\\workspace\\d_flichtenheld_core-jenkins-test\\<repo>@tmp\\jenkins-gitclient-ssh453049432504942179.key": invalid format
git@bitbucket.org: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2671)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:2096)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$500(CliGitAPIImpl.java:84)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:618)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$2.execute(CliGitAPIImpl.java:847)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:158)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:151)
	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
	at hudson.remoting.Request$2.run(Request.java:376)
	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:126)
	at java.lang.Thread.run(Thread.java:750)
```

### Anything else?

Key works on all agents when using the default credentials provider. Which is why I think this might be a bug in the plugin.

Not sure how to debug this further since the key is available as a temp file on the Windows agent only a very short time.

**Comment by chriskilding** [comment]
Is your Jenkins setup a distributed one (controller on one machine, agents that run the jobs on other machines), or all on one box?

**Comment by flichtenheld** [comment]
Controller is separate machine. All jobs run on distinct agents (both the Linux agents where the key works and the Windows agents where it doesn't)

**Comment by chriskilding** [comment]
Hmm. I have previously heard of a bug in distributed setups where, when a plugin consumes a credential (e.g. the git plugin, or slack plugin, or whatever), and the plugin does not snapshot the credential prior to transfer, the secret value ends up null when the job tries to use it. This doesn't happen with the Jenkins pipeline (`withCredentials` or `credentials` macros), which do take care to snapshot it beforehand.

If you can use Jenkins Pipeline in your setup, could you try adding

```groovy
node {
    withCredentials([sshUserPrivateKey(credentialsId: '<your SSH key id>', keyFileVariable: 'KEY', usernameVariable: 'KEY_USR')]) {
        // try to interact with $KEY or $KEY_USR some way, but you can't just print the value because this is masked in the build log
    }
}
```

before the git clone (or maybe just create a new job for this), and see what the result is?

I'm thinking that if something substantial is wrong, the credential won't bind or its value will effectively be null/empty string

**Comment by flichtenheld** [comment]
Those tests work reasonably fine. File is written correctly to the disk. But in other tests I just found out that this probably has nothing to do with Windows. I tried to use the key in sshagent (on Linux nodes) and that failed as well:

```text
[ssh-agent] Using credentials git (BitbucketSSH key)
[ssh-agent] Using credentials root (SCT SSH Key)
[ssh-agent] Looking for ssh-agent implementation...
[ssh-agent]   Exec ssh-agent (binary ssh-agent on a remote machine)
$ ssh-agent
SSH_AUTH_SOCK=/tmp/ssh-zmXXY9EUZ1wA/agent.4558
SSH_AGENT_PID=4560
Running ssh-add (command line suppressed)
Error loading key "/home/ubuntu/workspace/d_flichtenheld_core-jenkins-test@tmp/private_key_8107316556344895340.key": invalid format
```

Will do more tests with different keys and credential providers. Definitely seems to be more complicated than "doesn't work on Windows".

**Comment by chriskilding** [comment]
The AWS SSH private key class looks like this:

```java
public class AwsSshUserPrivateKey extends BaseStandardCredentials implements SSHUserPrivateKey {
    private static final Secret NO_PASSPHRASE = Secret.fromString("");

    private final Supplier<String> privateKey;
    private final String username;

    public AwsSshUserPrivateKey(String id, String description, Supplier<String> privateKey, String username) {
        super(id, description);
        this.privateKey = privateKey;
        this.username = username;
    }

    @NonNull
    @Deprecated
    @Override
    public String getPrivateKey() {
        return privateKey.get();
    }

    @Override
    public Secret getPassphrase() {
        return NO_PASSPHRASE;
    }

    @NonNull
    @Override
    public List<String> getPrivateKeys() {
        return Collections.singletonList(getPrivateKey());
    }

    @NonNull
    @Override
    public String getUsername() {
        return username;
    }

    // ...
}
```

And the credentials factory creates it by:

```java
String name;
String description;
String username;
<aws secrets manager client object> client;

new AwsSshUserPrivateKey(name, description, new StringSupplier(client, name), username)
```

And all the StringSupplier does is to get the secret-string from Secrets Manager.

So it's all pretty literal; whatever comes out of Secrets Manager, gets reflected in the SSH key object.

I'm wondering, however, whether encoding of the secret might make a difference. Perhaps one of the keys you tried had e.g. LF line breaks, while another had e.g. CRLF? And maybe this affects Windows?

**Comment by flichtenheld** [comment]
Sigh, after fidling around with the keys in secret manager I can't reproduce the issue currently. And during one of the tests I overrode the original value that caused the problems so I can't go back to try different things.

What I found out is that secretsmanager seems also doing some internal newline conversions when using --secret-string. When uploading a file with \r\n line ending with --secret-string the secret actually comes back with \n line endings. I know that at some point yesterday I managed to create a secret with \r\n line endings (which was not the problem, BTW), however now I'm not even able to reproduce that.

All a very long way of saying that at this point I will take the fact that it seems to work now and will only investigate further if I manage to break it again.

Closing for now.

Thanks for the help and sorry for the noise.
